test/test_replay.cpp

```diff
@@ -0,0 +1,57 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include <fusion.h>
+#include <kernel_cache.h>
+#include <ops/all_ops.h>
+#include <test/utils.h>
+#include <test/validator.h>
+#include <transform_replay.h>
+
+namespace nvfuser {
+
+using ReplayTest = NVFuserTest;
+
+TEST_F(ReplayTest, HorizontallyMergeTwoReshapes) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* in = makeContigConcreteTensor({4, 5});
+  TensorView* s0 = slice(in, {0, 0}, {4, 2});
+  TensorView* s1 = slice(in, {0, 2}, {4, 5});
+  TensorView* r0 = reshape(s0, {4, 2}, {2, 2, 2});
+  TensorView* r1 = reshape(s1, {4, 3}, {2, 2, 3});
+  TensorView* out = cat({r0, r1}, /*dim=*/-1);
+
+  std::vector<IterDomain*> new_out_root = IterDomain::clone(
+      TensorDomain::noReductions(in->getMaybeRFactorDomain()));
+  TensorView* new_out = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(new_out_root), *out->getDataType());
+  new_out->setDomain(
+      TransformReplay::fullSelfReplay(new_out->domain(), r0->domain()));
+  // FIXME: how about allocation domain and contiguity?
```

Comment:
I think that'll work. I'm fixing the `r0->getContiguity()` part -- should be easy. Code as is fails at https://github.com/samnordmann/Fuser/blob/28c36b6fe69dc67486cc08628a25269ce855aff2/csrc/ir/nodes.cpp#L3191, because before replay the maybeAllocationDomain is 2D not 3D. 

That being said, how would you go around the validation that I disabled? **Maybe** TransformReplay::fullSelfReplay is the wrong API and I should somehow publicize ReplayTransforms and use that instead?

Reply:
To be clear, the validation in fullSelfReplay fails at https://github.com/NVIDIA/Fuser/blob/71d9b555b401655c0d664b94e8abb7021ce29637/csrc/transform_replay.cpp#L205-L206 because r0's root domain has a rfactor product and IterDomain::clone removes rfactor product.

Reply:
I could also weaken that check 🤷